From e44d09160038c7cf661e23885a0997d54ca01752 Mon Sep 17 00:00:00 2001
Message-Id: <e44d09160038c7cf661e23885a0997d54ca01752.1417014698.git.erack@redhat.com>
In-Reply-To: <67f3ce3a9df2bc62db5602dd84975047c1137b92.1417014697.git.erack@redhat.com>
References: <67f3ce3a9df2bc62db5602dd84975047c1137b92.1417014697.git.erack@redhat.com>
From: Eike Rathke <erack@redhat.com>
Date: Mon, 17 Nov 2014 22:44:56 +0100
Subject: [PATCH 6/8] fdo#83765 do not update references in
 SortReorderByColumn() if disabled
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="------------erAck-patch-parts"

This is a multi-part message in MIME format.
--------------erAck-patch-parts
Content-Type: text/plain; charset=UTF-8; format=fixed
Content-Transfer-Encoding: 8bit


Similar to SortReorderByRow()

(cherry picked from commit 115a4b7ca36f65d93070d2e81048320d202e87a3)

Conflicts:
	sc/inc/column.hxx

Change-Id: I2d2fe243c91a663b14e5bd75ee30225d1b8073da
Reviewed-on: https://gerrit.libreoffice.org/13117
Reviewed-by: Kohei Yoshida <libreoffice@kohei.us>
Tested-by: Kohei Yoshida <libreoffice@kohei.us>
---
 sc/inc/column.hxx               | 12 +++++++-----
 sc/source/core/data/column4.cxx | 10 ++++++----
 sc/source/core/data/table3.cxx  |  3 ++-
 3 files changed, 15 insertions(+), 10 deletions(-)


--------------erAck-patch-parts
Content-Type: text/x-patch; name="0006-fdo-83765-do-not-update-references-in-SortReorderByC.patch"
Content-Transfer-Encoding: 8bit
Content-Disposition: attachment; filename="0006-fdo-83765-do-not-update-references-in-SortReorderByC.patch"

diff --git a/sc/inc/column.hxx b/sc/inc/column.hxx
index 66cd524..4bda39d 100644
--- a/sc/inc/column.hxx
+++ b/sc/inc/column.hxx
@@ -575,18 +575,20 @@ public:
 
     /**
      * Reset column position of formula cells within specified row range.
-     * Reference positions are also adjusted to reflect the new position so
-     * that the formula cells still reference the same cells or ranges after
-     * the the position change.  The position of a formula cell before the
-     * call is interpreted as the old position of that cell.
+     * If bUpdateRefs==true then reference positions are also adjusted to
+     * reflect the new position so that the formula cells still reference the
+     * same cells or ranges after the the position change.
+     * The position of a formula cell before the call is interpreted as the old
+     * position of that cell.
      *
      * Caller needs to ensure that no formula groups cross the top and bottom
      * row boundaries.
      *
      * @param nRow1 top row boundary
      * @param nRow2 bottom row boundary
+     * @param bUpdateRefs whether to adjust references
      */
-    void ResetFormulaCellPositions( SCROW nRow1, SCROW nRow2 );
+    void ResetFormulaCellPositions( SCROW nRow1, SCROW nRow2, bool bUpdateRefs );
 
     void SplitFormulaGroupByRelativeRef( const ScRange& rBoundRange );
 
diff --git a/sc/source/core/data/column4.cxx b/sc/source/core/data/column4.cxx
index 8b46e74..568b5b64 100644
--- a/sc/source/core/data/column4.cxx
+++ b/sc/source/core/data/column4.cxx
@@ -439,8 +439,9 @@ namespace {
 class FormulaColPosSetter
 {
     SCCOL mnCol;
+    bool  mbUpdateRefs;
 public:
-    FormulaColPosSetter( SCCOL nCol ) : mnCol(nCol) {}
+    FormulaColPosSetter( SCCOL nCol, bool bUpdateRefs ) : mnCol(nCol), mbUpdateRefs(bUpdateRefs) {}
 
     void operator() ( size_t nRow, ScFormulaCell* pCell )
     {
@@ -451,7 +452,8 @@ public:
             ScAddress aOldPos = pCell->aPos;
             pCell->aPos.SetCol(mnCol);
             pCell->aPos.SetRow(nRow);
-            pCell->GetCode()->AdjustReferenceOnMovedOrigin(aOldPos, pCell->aPos);
+            if (mbUpdateRefs)
+                pCell->GetCode()->AdjustReferenceOnMovedOrigin(aOldPos, pCell->aPos);
         }
         else
         {
@@ -463,9 +465,9 @@ public:
 
 }
 
-void ScColumn::ResetFormulaCellPositions( SCROW nRow1, SCROW nRow2 )
+void ScColumn::ResetFormulaCellPositions( SCROW nRow1, SCROW nRow2, bool bUpdateRefs )
 {
-    FormulaColPosSetter aFunc(nCol);
+    FormulaColPosSetter aFunc(nCol, bUpdateRefs);
     sc::ProcessFormula(maCells.begin(), maCells, nRow1, nRow2, aFunc);
 }
 
diff --git a/sc/source/core/data/table3.cxx b/sc/source/core/data/table3.cxx
index d646678..99d6239 100644
--- a/sc/source/core/data/table3.cxx
+++ b/sc/source/core/data/table3.cxx
@@ -716,8 +716,9 @@ void ScTable::SortReorderByColumn(
     }
 
     // Reset formula cell positions which became out-of-sync after column reordering.
+    bool bUpdateRefs = pArray->IsUpdateRefs();
     for (SCCOL nCol = nStart; nCol <= nLast; ++nCol)
-        aCol[nCol].ResetFormulaCellPositions(nRow1, nRow2);
+        aCol[nCol].ResetFormulaCellPositions(nRow1, nRow2, bUpdateRefs);
 
     // Set up column reorder map (for later broadcasting of reference updates).
     sc::ColRowReorderMapType aColMap;

--------------erAck-patch-parts--


